allennlp/models/semantic_parsing/atis/atis_semantic_parser.py

```diff
@@ -0,0 +1,592 @@
+                outputs['loss'] = self._decoder_trainer.decode(initial_state,
+                                                               self._decoder_step,
+                                                               (target_action_sequence, target_mask))['loss']
+            num_steps = self._max_decoding_steps
+            # This tells the state to start keeping track of debug info, which we'll pass along in
+            # our output dictionary.
+            initial_state.debug_info = [[] for _ in range(batch_size)]
+            best_final_states = self._beam_search.search(num_steps,
+                                                         initial_state,
+                                                         self._decoder_step,
+                                                         keep_final_unfinished_states=False)
+            outputs['best_action_sequence'] = []
+            outputs['debug_info'] = []
+            outputs['entities'] = []
+            outputs['logical_form'] = []
+            outputs['sql_queries'] = []
+            outputs['utterance'] = []
+            outputs['tokenized_utterance'] = []
+
+            for i in range(batch_size):
+                # Decoding may not have terminated with any completed logical forms, if `num_steps`
+                # isn't long enough (or if the model is not trained enough and gets into an
+                # infinite action loop).
+                if i in best_final_states:
+                    best_action_indices = best_final_states[i][0].action_history[0]
+
+                    action_strings = [action_mapping[(i, action_index)]
+                                      for action_index in best_action_indices]
+                    predicted_sql_query = self.action_sequence_to_sql(action_strings)
+
+                    if target_action_sequence is not None:
+                        # Use a Tensor, not a Variable, to avoid a memory leak.
+                        targets = target_action_sequence[i].data
+                        sequence_in_targets = 0
+                        sequence_in_targets = self._action_history_match(best_action_indices, targets)
+                        self._action_sequence_accuracy(sequence_in_targets)
+
+                        targets_list = [target.item() for target in targets[0]]
+                        similarity = difflib.SequenceMatcher(None, best_action_indices, targets_list)
+                        self._action_similarity(similarity.ratio())
+
+                    if sql_queries and sql_queries[i]:
+                        # Since the query might hang, we run in another process and kill it if it
+                        # takes too long.
+                        process = Process(target=self._sql_result_match,
+                                          args=(predicted_sql_query, sql_queries[i]))
+                        process.start()
+
+                        # If the query has not finished in 10 seconds then we will proceed.
```

Comment:
I'd probably give it a lot less than 10 seconds; the query should take milliseconds, I'd imagine.  You could also retry if you hang...  Not sure it's worth it.  You might consider including a metric that counts how many times you hang (you can prefix it with an underscore so it doesn't show up in tqdm).